`raft`: add `remake_learner_state` to `consensus` protocol [SC&R #5]

WillemKauf opened the pull request:
This PR is the 5th in the series of safe compaction & recovery [SC&R] PRs.

It adds the `remake_learner_state` primitive to the `raft/consensus` protocol for issuing `remake_partition()` requests (added in https://github.com/redpanda-data/redpanda/pull/26562) from a leader node on a learner node in the case that the leader detects the current round of recovery has entered a potentially divergent state.

This PR adds the types, classes and performs some plumbing through the `consensus` layer to wire everything up, but does not yet use any of the types defined here.

Future PRs in this series will tie everything together by using to be added state in the `log` and the now present `do_remake_learner_state()` commands present in the `consensus` object to ensure divergence free recovery.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes


* none

WillemKauf committed:
`raft`: add `consensus::truncate_state()` function

NFC, adds a new function `truncate_state()` to replace some code
which will be re-used in a later commit.

WillemKauf committed:
`raft`: add `_recovery_resets` to `probe`

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1a0e5c0f-0234-49c8-8277-830a7550d846-->
<!--TEST RESULTS-->
<!--6f7df8ad-8fe7-4d86-8512-8593f50bacb9-->
<!--TEST RESULTS-->
<!--bfa7abca-1183-43a2-94b5-2d74127dee9b-->
<!--TEST RESULTS-->
<!--053b2201-783f-4a38-bf52-a0f580155e7c-->
<!--TEST RESULTS-->
<!--9cd85e6d-7b70-426f-9ed2-d213ad655505-->
<!--TEST RESULTS-->
<!--d21d2009-69ee-4319-b689-88df855676ce-->
<!--TEST RESULTS-->
<!--ccabf926-8758-498c-80c2-57c4919bd2a6-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68831">build#68831</a></summary>


| test_class                   | test_method                             | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:----------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves | {"cleanup_policy": "compact,delete"}                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68831#0197fa40-82dd-442a-9915-827cfd128cb1 | FLAKY         | 18/21    | upstream reliability is '95.29411764705881'. current run reliability is '85.71428571428571'. drift is 9.57983 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest     | test_node_operations                    | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68831#0197fa40-82dc-4ef2-8b0d-61a9ff2371c7 | FLAKY         | 20/21    | upstream reliability is '98.90510948905109'. current run reliability is '95.23809523809523'. drift is 3.66701 and the allowed drift is set to 50. The test should PASS |
| TestTieredStoragePause       | test_safe_pause_resume                  | {"allow_gaps_cluster_level": true, "allow_gaps_topic_level": null}                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68831#0197fa40-82e1-4f6c-9962-dded587895c2 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68917">build#68917</a></summary>


| test_class               | test_method            | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-----------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations   | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68917#019809a6-9710-46a0-997a-7bbd3e537c0e | FLAKY         | 20/21    | upstream reliability is '98.90710382513662'. current run reliability is '95.23809523809523'. drift is 3.66901 and the allowed drift is set to 50. The test should PASS |
| TestTieredStoragePause   | test_safe_pause_resume | {"allow_gaps_cluster_level": false, "allow_gaps_topic_level": true}                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68917#019809a6-9707-4926-b3c7-e7edd6442daa | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68942">build#68942</a></summary>


| test_class       | test_method              | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------|:-------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeE2ETests | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68942#01980ac6-e12a-4033-9a92-868c90e4bf99 | FLAKY         | 19/21    | upstream reliability is '97.68339768339769'. current run reliability is '90.47619047619048'. drift is 7.20721 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69404">build#69404</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69404#01982ff2-44e8-4049-b5a8-602c74c693f5 | FLAKY         | 18/21    | upstream reliability is '96.61016949152543'. current run reliability is '85.71428571428571'. drift is 10.89588 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69448">build#69448</a></summary>


| test_class                     | test_method                       | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-------------------------------|:----------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| CompactionTermRollRecoveryTest | test_compact_term_rolled_recovery |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69448#01983317-d339-4333-b8ec-88d142e958f5 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
| CertificateRevocationTest      | test_kafka                        |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69448#01983316-10d5-43f8-ac45-a2914399ae14 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest       | test_node_operations              | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69448#01983317-d335-43fd-8b8c-254cf8646947 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69481">build#69481</a></summary>


| test_class               | test_method                        | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-----------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest    | test_migrated_topic_data_integrity | {"params": {"cancellation": {"dir": "in", "stage": "prepared"}, "use_alias": true}, "transfer_leadership": true}                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69481#0198350e-d45a-4c78-bcd0-54b3d4e0f961 | FLAKY         | 20/21    | upstream reliability is '99.72752043596729'. current run reliability is '95.23809523809523'. drift is 4.48943 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations               | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69481#019834fb-a43f-4591-be86-96a6e87ad0d9 | FLAKY         | 19/21    | upstream reliability is '99.58677685950413'. current run reliability is '90.47619047619048'. drift is 9.11059 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69547">build#69547</a></summary>


| test_class                  | test_method                               | test_arguments                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------|:------------------------------------------|:--------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| JavaCompressionTest         | test_upgrade_java_compression             | {"compression_type": "gzip"}                                        | integration | https://buildkite.com/redpanda/redpanda/builds/69547#01983872-3346-4846-87ac-8caa1dd26ab3 | FLAKY         | 20/21    | upstream reliability is '95.1048951048951'. current run reliability is '95.23809523809523'. drift is -0.1332 and the allowed drift is set to 50. The test should PASS   |
| JavaCompressionTest         | test_upgrade_java_compression             | {"compression_type": "gzip"}                                        | integration | https://buildkite.com/redpanda/redpanda/builds/69547#01983873-3e91-4a70-9d64-8466d94f7ad0 | FLAKY         | 20/21    | upstream reliability is '95.1048951048951'. current run reliability is '95.23809523809523'. drift is -0.1332 and the allowed drift is set to 50. The test should PASS   |
| JavaCompressionTest         | test_upgrade_java_compression             | {"compression_type": "lz4"}                                         | integration | https://buildkite.com/redpanda/redpanda/builds/69547#01983872-3347-4a5e-b78b-89d993a282a4 | FLAKY         | 20/21    | upstream reliability is '92.41379310344827'. current run reliability is '95.23809523809523'. drift is -2.8243 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest         | test_upgrade_java_compression             | {"compression_type": "lz4"}                                         | integration | https://buildkite.com/redpanda/redpanda/builds/69547#01983873-3e93-44d1-9cf9-43545f6b63fa | FLAKY         | 20/21    | upstream reliability is '92.41379310344827'. current run reliability is '95.23809523809523'. drift is -2.8243 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest         | test_upgrade_java_compression             | {"compression_type": "snappy"}                                      | integration | https://buildkite.com/redpanda/redpanda/builds/69547#01983872-3349-45d9-ab2a-c53b045e8837 | FLAKY         | 20/21    | upstream reliability is '92.46575342465754'. current run reliability is '95.23809523809523'. drift is -2.77234 and the allowed drift is set to 50. The test should PASS |
| JavaCompressionTest         | test_upgrade_java_compression             | {"compression_type": "snappy"}                                      | integration | https://buildkite.com/redpanda/redpanda/builds/69547#01983873-3e93-495e-a695-2afbd0b9e69a | FLAKY         | 20/21    | upstream reliability is '92.46575342465754'. current run reliability is '95.23809523809523'. drift is -2.77234 and the allowed drift is set to 50. The test should PASS |
| JavaCompressionTest         | test_upgrade_java_compression             | {"compression_type": "zstd"}                                        | integration | https://buildkite.com/redpanda/redpanda/builds/69547#01983872-334a-4e9e-9161-d3b6e6b0246c | FLAKY         | 20/21    | upstream reliability is '92.3076923076923'. current run reliability is '95.23809523809523'. drift is -2.9304 and the allowed drift is set to 50. The test should PASS   |
| JavaCompressionTest         | test_upgrade_java_compression             | {"compression_type": "zstd"}                                        | integration | https://buildkite.com/redpanda/redpanda/builds/69547#01983873-3e94-4cff-b635-20e76dd86dd1 | FLAKY         | 20/21    | upstream reliability is '92.3076923076923'. current run reliability is '95.23809523809523'. drift is -2.9304 and the allowed drift is set to 50. The test should PASS   |
| IdempotencySnapshotDelivery | test_recovery_after_snapshot_is_delivered | null                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/69547#01983873-3e91-4a70-9d64-8466d94f7ad0 | FLAKY         | 20/21    | upstream reliability is '98.29545454545455'. current run reliability is '95.23809523809523'. drift is 3.05736 and the allowed drift is set to 50. The test should PASS  |
| TestTieredStoragePause      | test_safe_pause_resume                    | {"allow_gaps_cluster_level": false, "allow_gaps_topic_level": true} | integration | https://buildkite.com/redpanda/redpanda/builds/69547#01983872-3349-45d9-ab2a-c53b045e8837 | FLAKY         | 20/21    | upstream reliability is '98.42931937172776'. current run reliability is '95.23809523809523'. drift is 3.19122 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

mmaslankaprv commented on src/v/raft/recovery_rpc.json:
```diff
@@ -0,0 +1,14 @@
+    ],
+    "methods": [
+        {
+            "name": "reset_learner_state",
```
can we use `remake` name here to minimise the number of names we use for the same operation ?

WillemKauf replied:
Replaced all instances of `reset_learner_*` with `remake_learner`.

mmaslankaprv commented on src/v/raft/recovery_rpc_types.h:
```diff
@@ -0,0 +1,66 @@
+
+    auto serde_fields() { return std::tie(ntp); }
+
+    model::ntp ntp;
```
Raft uses `group_id` to identify the raft group instance, should we use the same here ?

WillemKauf replied:
I at first had this as `group_id` instead of `ntp`- however, since we are issuing `remake_partition()` commands using this RPC, it felt more "direct" to use the `ntp` instead of raft group.

If you feel strongly about this, I _think_ I can change it back, but at some point there will have to be a `group_id` -> `ntp` matching to issue a `remake_partition()` command with the `controller_api`.

andrwng replied:
+1, if the caller here is going to be Raft, it seems like we should operate in the Raft space?

Also, it feels a bit off to me to not have partition revision here in some form. Has that already been considered and deemed safe?

WillemKauf replied:
> if the caller here is going to be Raft, it seems like we should operate in the Raft space?

Happy to make the change if necessary, `controller_api` can invoke `consensus_for(group_id)` or `partition_for(group_id)` using its held `partition_manager` to grab the `ntp` from the `group_id` when invoking `remake_partition(ntp)`.

> Also, it feels a bit off to me to not have partition revision here in some form. Has that already been considered and deemed safe?

The `recovery_stm` doesn't seem to concern itself with the `partition` revision at all during a recovery, I'm not sure if its necessary here or not?

mmaslankaprv replied:
I would go with the `group_id` it is a unique identifier that changes independently of the `ntp`

mmaslankaprv commented on src/v/raft/recovery_rpc_handler.h:
```diff
@@ -0,0 +1,43 @@
+    recovery_rpc_handler(
+      ss::scheduling_group sg,
+      ss::smp_service_group ssg,
+      ss::sharded<cluster::controller_api>& api)
```
i am not a fan of Raft depending on `cluster` module, can we figure out the way to do it differently ?

WillemKauf replied:
I would love a way to do this differently because I agree that `raft` typically does not depend on `cluster`, however this feels required because we are issuing `remake_partition()` commands through the `controller_api` to the `controller_backend`, which is where `partition` remaking must happen. 

To alleviate some of the pain here, the `raft_recovery_rpc_handler` is its own bazel cc library which depends on both `raft` and `cluster`, so there is no major dependency leakage/side effects. 

Is that acceptable to you, or do you see any way to circumvent the dependency upon `cluster` (because I do not)?

andrwng replied:
Maybe it makes sense to wrap `controller_api` in some Raft abstract class like `raft::remake_learner_dispatcher` or something that the controller can implement with `remake_partition`?

mmaslankaprv replied:
love that idea from Andrew

mmaslankaprv commented on src/v/raft/tests/raft_fixture.cc:
```diff
@@ -476,6 +481,8 @@ raft_node_instance::initialise(std::vector<raft::vnode> initial_nodes) {
       config::mock_binding<bool>(_enable_longest_log_detection),
       consensus_client_protocol(_buffered_protocol),
+      recovery_client_protocol(
+        ss::make_shared<recovery_rpc_client>(_id, _connections)),
```
The whole beauty of the raft fixture was that it did not depend on network connections, maybe we can make this API part of Raft protocol, this way we can abstract away connections and still use in memory transport in the tests

WillemKauf replied:
I had some changes from my main SC&R branch in a later commit that changes this instead to `in_memory_recovery_client`, I have pulled those changes into the fixture here. Let me know if that's better.

andrwng commented on src/v/redpanda/application.cc:
```diff
@@ -3260,6 +3261,12 @@ void application::start_runtime_services(
 
+          runtime_services.push_back(
+            std::make_unique<raft::recovery_rpc_handler>(
+              sched_groups.raft_heartbeats(),
```
Curious, does `raft_recv_sg()` make sense here?

WillemKauf replied:
I thought it was the closest scheduling group that made sense, and the hope is that this RPC is already incredibly infrequent anyways. Happy to change it if there are any strong opinions.

andrwng replied:
Ah oops, i assumed recv was "recovery" here.

I guess generally I wonder, as rare was we expect it to be, if it's more undesirable for the logic in `remake_partition` to contend with heartbeats (critical for stability, generally lightweight, high priority), vs other Raft traffic

WillemKauf replied:
@mmaslankaprv Please let us know if you have an opinion on `raft_recv_sg()` vs `raft_heartbeats()` for the scheduling group of this rpc service.

mmaslankaprv commented on src/v/raft/recovery_rpc_types.h:
```diff
@@ -0,0 +1,66 @@
+
+namespace raft {
+
+struct remake_learner_state_request
```
Other `raft` requests/replies have the `source` and `target` vnode_ids, should we consider doing the same here, i think we should. Old partition replicas (from isolated or lagging nodes) may send requests to current instances, by validating the `target_vnode` we are certain that the request is sent to the target node

WillemKauf committed:
`raft`: add `_remake_cb` and `set/clear` functions to `group_manager`

The request to remake a `partition` will bubble up from `raft` into
the `cluster` layer, rather than a direct line of communication from an
intermediate object which depends on both `raft` and `cluster` libraries.

The idea is that the `consensus` will invoke a callback registered by
the `group_manager`, which in turn will invoke a callback registered by
the `controller` between the `group_manager` and `controller_api` to
invoke a `remake_partition()` command.

This commit adds the `_remake_cb` as well as the `set/clear_remake_cb()`
functions. Since the lifetime of the `group_manager` and `controller_backend/api`
are not tied, `clear_remake_cb()` needs to be used to ensure no calls to
stopped services are made (in the situation that we receive a `remake_partition()`
request in the middle of shutdown).

WillemKauf committed:
`cluster`: add `remake_cb` functions to `controller`

Here, we set the necessary callbacks between the `group_manager`
and `controller_api` in order to allow a `consensus` object to
bubble up a request to `controller_backend::remake_partition()` from
the `raft` layer.

This callback is set in `controller::start()` _after_ the `controller_backend`
and `controller_api` are started, and cleared in `controller::stop()`
_before_ the `controller_backend` and `controller_api` are stopped.

WillemKauf committed:
`cluster`: use `raft::group_id` instead of `model::ntp` in `remake_partition()`

Per feedback, this is altered to use the `raft::group_id` from a
`remake_learner_state_request` instead of a raw `ntp`. The `controller_api`
 now makes an extra request to the `partition_manager` to get the appropriate
`ntp` from a `group_id`.

WillemKauf committed:
`raft`: add `remake_cb_t` to `consensus` constructor

Adds the callback for `consensus -> group_manager` communication in
order to invoke a call into `controller_api::remake_partition()`.

WillemKauf committed:
`raft`: add `consensus::{do}_remake_learner_state()`

The functions invoked on the leader (send) & learner (recv) side to clear
local state on the learner when it is deemed necessary.

WillemKauf committed:
`raft`: add `remake_learner_state_request` to `raft` protocol

Adds the `remake_learner_state` request/reply protocol to the `raft`
rpc handlers and provides an implementation for `rpc_client_protocol`
as well as `in_memory_test_protocol`.

WillemKauf committed:
`raft`: add `remake_learner_state_request/reply`

WillemKauf committed:
`raft`: add `probe::get_log_truncations()`

For use in testing.

WillemKauf commented:
I rewrote 99% of the PR after discussions with @mmaslankaprv. Please LMKWYT.

andrwng commented on src/v/raft/group_manager.h:
```diff
@@ -99,6 +99,18 @@ class group_manager {
+
+    // Lifetime of `group_manager` & `controller_backend/api` is not tied. We
+    // need to clear this callback to avoid calls to stopped services.
+    void clear_remake_cb() { _remake_cb = nullptr; }
```
Do we need to hold a gate to ensure the callback isn't in flight when this is destructed?

WillemKauf replied:
Hmm yeah. This probably needs some form of concurrency control, thanks for pointing out the oversight.

WillemKauf replied:
Added some concurrency control with a new `_remake_cb_gate` in `group_manager.h`.

andrwng commented on src/v/raft/consensus.cc:
```diff
@@ -4290,4 +4290,80 @@ size_t consensus::bytes_to_deliver_to_learners() const {
+      .term = _term};
+    vlog(_ctxlog.debug, "Issuing remake group request {}", req);
+    static constexpr auto timeout = 600s;
+    result<remake_learner_state_reply> reply
+      = co_await _client_protocol.remake_learner_state(
+        target.id(), req, rpc::client_opts(timeout));
```
Naive question since I haven't looked into the remake_partition implementation -- does this also reset the consensus object? Or just the partition and log state?

WillemKauf replied:
We call into `partition_manager::remove()` and `remove_persistent_state()` from `controller_backend::remake_partition()`:

https://github.com/redpanda-data/redpanda/blob/6a6af3c379cb14555a2f1455289dad3bd11f9c6d/src/v/cluster/controller_backend.cc#L2130-L2140

This calls into `group_manager::remove()` as well as removing `partition/log` state.

https://github.com/redpanda-data/redpanda/blob/6a6af3c379cb14555a2f1455289dad3bd11f9c6d/src/v/cluster/partition_manager.cc#L438-L450

mmaslankaprv commented on src/v/raft/consensus.cc:
```diff
@@ -4287,4 +4290,80 @@ size_t consensus::bytes_to_deliver_to_learners() const {
+      .target_node_id = target,
+      .group = _group,
+      .term = _term};
+    vlog(_ctxlog.debug, "Issuing remake group request {}", req);
```
lets log it with info level, it is very rare event and would be great to know about it

WillemKauf replied:
Changed to `info` level

mmaslankaprv commented on src/v/raft/consensus.cc:
```diff
@@ -4287,4 +4290,80 @@ size_t consensus::bytes_to_deliver_to_learners() const {
+
+        std::optional<raft::errc> err;
+        // Perform validation of request under _op_lock
+        if (req.term != _term) {
+            err = raft::errc::leadership_transfer_in_progress;
```
can we retrun after each check so that the error is not lost

mmaslankaprv replied:
return `not_leader` error

WillemKauf replied:
The error shouldn't be lost:

```cpp
std::optional<raft::errc> err;
// Perform validation of request under _op_lock
if (req.term != _term) {
    err = raft::errc::leadership_transfer_in_progress;
}
if (
  !_leader_id.has_value() || req.source_node() != _leader_id.value()) {
    err = raft::errc::leadership_transfer_in_progress;
}
if (req.target_node() != _self) {
    err = raft::errc::invalid_target_node;
}
if (req.group != _group) {
    err = raft::errc::group_not_exists;
}

if (err.has_value()) {
    reply.success = is_success::no;
    vlog(
      _ctxlog.warn,
      "Unable to process remake group request {}, raft::errc {}",
      req,
      err.value());
}
```

We log exactly what it is, though the `reply` only contains a boolean `yes/no` response.

Does that make sense?

mmaslankaprv replied:
but it can be replaced by the subsequent if condition

WillemKauf replied:
Replaced this with a `lambda` function and early returns

mmaslankaprv commented on src/v/raft/consensus.cc:
```diff
@@ -4287,4 +4290,80 @@ size_t consensus::bytes_to_deliver_to_learners() const {
+            err = raft::errc::leadership_transfer_in_progress;
+        }
+        if (
+          !_leader_id.has_value() || req.source_node() != _leader_id.value()) {
```
`req.source_node() != _leader_id` shold be enough

WillemKauf replied:
Fixed

mmaslankaprv commented on src/v/raft/consensus.cc:
```diff
@@ -4287,4 +4290,80 @@ size_t consensus::bytes_to_deliver_to_learners() const {
+}
+
+ss::future<remake_learner_state_reply>
+consensus::do_remake_learner_state(remake_learner_state_request req) {
```
please add a check validating if node receiving the request is a learner

WillemKauf replied:
Added

mmaslankaprv commented:
overall looks great, just few comments, the most important one is about the `learner` check

WillemKauf committed:
`raft`: add `consensus::is_learner()`

A utility function to check whether this node is a learner or not,
based on the fact that a learner will not yet be a voter.

WillemKauf committed:
`raft`: add `errc::not_learner`

WillemKauf committed:
`raft`: add `reset_learner_state` fixture test

WillemKauf commented on src/v/raft/consensus.cc:
```diff
@@ -4290,4 +4290,85 @@ size_t consensus::bytes_to_deliver_to_learners() const {
+      .group = _group,
+      .term = _term};
+    vlog(_ctxlog.info, "Issuing remake group request {}", req);
+    static constexpr auto timeout = 600s;
```
@mmaslankaprv not sure how you feel about this hardcoded timeout, and whether it should be tweakable or not.

mmaslankaprv replied:
i think it should be smaller f.e. 10 seconds

WillemKauf replied:
I assumed that this remake operation could potentially be pretty heavy on the learner (what if we were talking about ~100s of GBs on the learner needing to be removed?), which is why I felt `600s` was appropriate, do you think `10s` is closer to the appropriate order of magnitude?

mmaslankaprv replied:
if i understand the flow correctly the RPC will not wait for deletion to finish ?

WillemKauf replied:
>if i understand the flow correctly the RPC will not wait for deletion to finish ?

This is currently true, I did have a future commit in another branch that changed that within the `controller_api`: https://github.com/redpanda-data/redpanda/commit/bf51b32dca0f24abeac6ece22e8403a7a0f58f24

Maybe I should pull it into this PR.

WillemKauf replied:
Changed this timeout to `10s`. Leaving the RPC as an asynchronous `remake_partition()` command on the target node.

WillemKauf commented on src/v/raft/consensus.cc:
```diff
@@ -4290,4 +4290,85 @@ size_t consensus::bytes_to_deliver_to_learners() const {
+            if (req.target_node() != _self) {
+                return raft::errc::invalid_target_node;
+            }
+            if (!is_learner()) {
```
I worry that this `if()` is going to be a blocker for removable transactional control batches.

@mmaslankaprv and I discussed how remaking the partition for a node that is currently a voter is dangerous for data loss, but if this voter is in recovery it is still subject to the same divergence we are trying to avoid with this whole approach. We discussed that to handle this case, we would have to perform raft group reconfiguration:

> for example assuming replica 3 is the lagging one:
> go from 1,2,3 -> 1,2 -> 1,2,3
> this is the only safe way of doing that
> but in the period of 1,2 it is related with availability loss

@bharathv I am not sure if you have any opinions here, I think we need to be airtight in all situations with recovery before we can consider removing transactional control batches.

mmaslankaprv replied:
I think we can also demote the voter to learner, then remake the learner and promote it to voter back, but it still impacts the availability

WillemKauf replied:
Should this be done in all cases when `dispatch_recovery()` is called?

bharathv replied:
> and I discussed how remaking the partition for a node that is currently a voter is dangerous for data loss, b

for my understanding, whats the problematic scenario here, this is a voter but not the longest log (since it is recovering), so there is a majority quorum not including the voter we trying to reset.

WillemKauf commented:
Adding two new commits on the tail end here per conversations about issuing `remake_partition()` commands on existing voters in a quorum.

I don't know if this is safe or not, hoping some `raft` SMEs can review and let me know what they think.

WillemKauf commented:
Force push:
* Remove last two commits per discussions with replication team
* Change RPC timeout in `consensus::remake_learner_state()` to `10s` from `600s`.

WillemKauf commented:
We still do not yet have a solution to divergence free recovery with voter nodes in a Raft quorum. This is left as future work, the current PR will only satisfy divergence free recovery for learner nodes when it is hooked up in a future PR.
